exchanges/stream/websocket_test.go

```diff
@@ -0,0 +1,968 @@
+	}
+
+	err = ws.UnsubscribeChannels(subs)
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestResubscribe(t *testing.T) {
+	ws := *New()
+	err := ws.Setup(defaultSetup)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	fnSub := func(subs []ChannelSubscription) error {
+		ws.AddSuccessfulSubscriptions(subs...)
+		return nil
+	}
+	fnUnsub := func(unsubs []ChannelSubscription) error {
+		ws.RemoveSuccessfulUnsubscriptions(unsubs...)
+		return nil
+	}
+	ws.Subscriber = fnSub
+	ws.Unsubscriber = fnUnsub
+
+	channel := []ChannelSubscription{{Channel: "resubTest"}}
+	err = ws.ResubscribeToChannel(&channel[0])
+	if err == nil {
+		t.Fatal("error cannot be nil")
+	}
+
+	err = ws.SubscribeToChannels(channel)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	err = ws.ResubscribeToChannel(&channel[0])
+	if err != nil {
+		t.Fatal("error cannot be nil")
+	}
+}
+
+// TestConnectionMonitorNoConnection logic test
+func TestConnectionMonitorNoConnection(t *testing.T) {
+	ws := *New()
+	ws.DataHandler = make(chan interface{}, 1)
+	ws.ShutdownC = make(chan struct{}, 1)
+	ws.exchangeName = "hello"
+	ws.trafficTimeout = 1
+	go ws.connectionMonitor()
+	time.Sleep(time.Second)
+	if ws.IsConnectionMonitorRunning() {
+		t.Fatal("Should have exited")
+	}
+}
+
+// TestSliceCopyDoesntImpactBoth logic test
+func TestGetSubscriptions(t *testing.T) {
+	w := Websocket{
+		subscriptions: []ChannelSubscription{
+			{
+				Channel: "hello3",
+			},
+		},
+	}
+	if !strings.EqualFold("hello3", w.GetSubscriptions()[0].Channel) {
+		t.Error("Subscriptions was not copied properly")
+	}
+}
+
+// TestSetCanUseAuthenticatedEndpoints logic test
+func TestSetCanUseAuthenticatedEndpoints(t *testing.T) {
+	ws := *New()
+	result := ws.CanUseAuthenticatedEndpoints()
+	if result {
+		t.Error("expected `canUseAuthenticatedEndpoints` to be false")
+	}
+	ws.SetCanUseAuthenticatedEndpoints(true)
+	result = ws.CanUseAuthenticatedEndpoints()
+	if !result {
+		t.Error("expected `canUseAuthenticatedEndpoints` to be true")
+	}
+}
+
+const (
+	websocketTestURL  = "wss://www.bitmex.com/realtime"
+	returnResponseURL = "wss://ws.kraken.com"
+	useProxyTests     = false                     // Disabled by default. Freely available proxy servers that work all the time are difficult to find
+	proxyURL          = "http://212.186.171.4:80" // Replace with a usable proxy server
+)
+
+var dialer websocket.Dialer
+
+type testStruct struct {
+	Error error
+	WC    WebsocketConnection
+}
+
+type testRequest struct {
+	Event        string          `json:"event"`
+	RequestID    int64           `json:"reqid,omitempty"`
+	Pairs        []string        `json:"pair"`
+	Subscription testRequestData `json:"subscription,omitempty"`
+}
+
+// testRequestData contains details on WS channel
+type testRequestData struct {
+	Name     string `json:"name,omitempty"`
+	Interval int64  `json:"interval,omitempty"`
+	Depth    int64  `json:"depth,omitempty"`
+}
+
+type testResponse struct {
+	RequestID int64 `json:"reqid,omitempty"`
+}
+
```

Comment:
Can we reorder these to the top just to keep it clearer?

also returnResponseURL is not used currently